### DISCUSSION TITLE
Why is `googletest` not a git submodule of the repository?

### DISCUSSION THREAD

**Discussion by ThibFrgsGmz** [nasa/fprime#1819] [Q&A] [answered]
Hello,

Could you tell me the main reasons why `googletest` is not specified as a submodule of the project please?

In my opinion, since it is an essential dependency, it makes more sense to have it in a submodule so that we have the source of `googletest` with the repo clone.

Currently, the installation of `googletest` is done later in the project when the developer wants to implement unit tests.

**Comment by LeStarch** [comment, marked as answer]
Good question!  

This was originally implemented to remove the use of git submodules on the core repository. Git submodule usage can be non-intuitive and can cause hard to debug errors if used in an unexpected way (recursive submodule use).  Thus to remove this issue we integrated the tool deployment as part of the build instead.

**Reply by bocchino** [reply]
> Git submodule usage can be non-intuitive

Here's one reason for this: if a git submodule pointer changes on a branch of the remote repo, say `devel`, and you pull down `devel`, things do not work as you would expect. Normally if you check out `devel` and do `git pull`, everything is in sync with the head of `devel` on the remote repo. Submodules instead remain in their local state until you run `git submodule update` on them. While this behavior must be useful to some developers, in my experience it is the opposite of what is wanted. For users who don't know about git submodules, it's confounding. The default behavior is to let the local submodule go quietly out of date without any warning except the "modified" message when you run `git status`, which is confusing. And even if you notice the status message, there's no hint about how to fix it. You have to ask someone or search for info about `git submodule update` on google. Doing the obvious thing (e.g., `git pull` or `git checkout`) doesn't fix the problem.

**Reply by bocchino** [reply]
I think the upshot of this is that if we made gtest a submodule, we would pretty much guarantee that some F Prime users would eventually use the wrong version of gtest without knowing it unless it caused a problem, at which point they would have no idea how to fix it.

**Reply by bocchino** [reply]
Also, as Michael pointed out, since F Prime is already usually a submodule of a project, making gtest a submodule of F Prime would introduce a submodule of a submodule, which is even more confusing than what I described above.

**Reply by astroesteban** [reply]
I've used repos before that have like 3 levels of nested submodules. Those are not fun problems to debug. Even for a single submodule. Special shout-out to Git LFS which is also a huge pain to debug.

**Reply by ThibFrgsGmz** [reply]
OK, I more or less understand your point of view. And indeed, I find it difficult to manage several levels of sub-modules.

On the other hand, is it possible to better automate the download of `GTest`?

Currently, we have to go through `fprime-util generate --ut` to trigger the `GTest` download.

This is not very practical because in principle once you do the `pip install` you expect all the dependencies of the project to be installed and this is not the case (same thing for `fprime- visual`/`layout`).

In industrial environments such as space/defense we often don't have access to the outside. Depending on a company's security policies, it is possible to exceptionally open the flows to the outside while installing and then close them as soon as possible. So when we miss a dependency, it's quite annoying (I had this case several times for FPrime with `GTest`/`FPL`/`FPV`).

**Reply by LeStarch** [reply]
We could pull the GTest download into the PIP install. That way it would be in the "same step" as the other installations.  It would not be so far out of band as we do this for FPP....it would be just another minor transgression against PIP.

**Reply by LeStarch** [reply]
Alternatively we could have users just clone it alongside F´. This is more for users to remember, but less for us to manage.
